# Recommender
Marcus Millender

Recommender Homework
* Step 1:  Download the recommender_homework.ipynb file from my git
* Step 2:  Follow the instructions contained in that notebook
* Step 3:  Submit your answers by sending a link to your github submission.

In [1]:
import pandas as pd
import numpy as np

In [2]:
#download the class movie reviews from here:
#https://docs.google.com/spreadsheets/d/17rCJzmWxqvAu9rkpkgt4ToccIlY4A1Ffuu1W9X3B8Ag/
#then read that in as a pandas dataframe

In [3]:
M = pd.read_csv("Movie Reviews.csv", index_col='Name')

In [4]:
def pearson(s1, s2):
    """Take two pd.Series objects and return a pearson correlation."""
    s1_c = s1 - s1.mean()
    s2_c = s2 - s2.mean()
    return np.sum(s1_c * s2_c) / np.sqrt(np.sum(s1_c ** 2) * np.sum(s2_c ** 2))

In [5]:
def get_recs(movie_name, M, num):

    import numpy as np
    reviews = []
    for title in M.columns:
        if title == movie_name:
            continue
        cor = pearson(M[movie_name], M[title])
        if np.isnan(cor):
            continue
        else:
            reviews.append((title, cor))
    
    reviews.sort(key=lambda tup: tup[1], reverse=True)
    return reviews[:num]

    

### Question 1:  What movie is most Similar to 'The fault in our stars' (60 pts)


In [6]:
recs = get_recs('The Fault in Our Stars', M, 1)

In [7]:
recs

[('Malificent', 0.19749425916269375)]

#### Answer 1: 
Based on the reviews, Malificent is most similar to 'The fault in our stars'

### Question 2:  Which movie(s) would you most like to see, based on your classmates experience? (40 pts)

In [8]:
#HINTS...

#the movies I've seen
my_ratings = M.ix['Marcus Millender'].dropna()
#the movie I've rated the highest
my_ratings.sort(ascending=False)
#the name of the movie I've rated the highest
fav_movie = my_ratings.index[0]

C:\Users\millm_000\Anaconda3\lib\site-packages\ipykernel\__main__.py:6: FutureWarning: sort is deprecated, use sort_values(inplace=True) for INPLACE sorting


In [9]:
my_ratings

Godzilla                                 5.0
Big Hero 6                               5.0
Malificent                               5.0
Guardians of the Galaxy                  5.0
300: Rise of an Empire                   4.0
How to Train your Dragon 2               4.0
Interstellar                             4.0
The Hunger Games: Mockingjay - Part 1    4.0
American Sniper                          4.0
Divergent                                3.0
Name: Marcus Millender, dtype: float64

In [10]:
fav_movie
# Actually, my favorite is Guardians of the Galaxy

'Godzilla'

In [11]:
recs = get_recs(fav_movie, M, 3)

In [12]:
recs

[('Transformers', 0.29963535888671367),
 ('Malificent', 0.28957046964673355),
 ('Interstellar', 0.18186793275523092)]

#### Answer 2: 

Based on my classmates experience, the top three movies I would most like to see are Trasformers, Malificent, and Interstellar.

Tranformers
   + Though I generally don't like the Transformers films after the first one, the second one killed it for me.  So I have not seen this or any other Transformers movie after the second. Maybe they got better?

Malificent
+ I can say that I enjoyed Malificent.  That is a good pick.

Interstellar
+ I did enjoy Interstellar as well, I could see it again.

## Question 3: Bonus Question...  
#### For all the movies you haven't seen, can you predict how you'd rate them using your the class reviews? (10 pts)

In [13]:
# List the movies I haven't seen(True)
have_not_seen = M.ix['Marcus Millender'].isnull()

In [14]:
have_not_seen

American Sniper                          False
The Hunger Games: Mockingjay - Part 1    False
Guardians of the Galaxy                  False
The Lego Movie                            True
The Hobbit                                True
Transformers                              True
Malificent                               False
Big Hero 6                               False
Godzilla                                 False
Interstellar                             False
How to Train your Dragon 2               False
Gone Girl                                 True
Divergent                                False
The Fault in Our Stars                    True
Unbroken                                  True
300: Rise of an Empire                   False
Name: Marcus Millender, dtype: bool

#### From the list of movies I have not seen 6 of the movies:

* The Lego Movie
* The Hobbit
* Gone Girl
* The Fault in Our Stars
* Unbroken
* Transformers

In [15]:
my_ratings

Godzilla                                 5.0
Big Hero 6                               5.0
Malificent                               5.0
Guardians of the Galaxy                  5.0
300: Rise of an Empire                   4.0
How to Train your Dragon 2               4.0
Interstellar                             4.0
The Hunger Games: Mockingjay - Part 1    4.0
American Sniper                          4.0
Divergent                                3.0
Name: Marcus Millender, dtype: float64

#### Answer 3: How I will predict the score
From the class reviews, for the movies I haven't seen, I would predict the ratings of those movies by adding the (class average of movie 'X' I have seen) to the {[correlation of (a movie 'Y' I haven't seen) and (the move 'X' I have seen)] multiplied by the score of movie 'X' I have seen}

+ The movie I will compare is '300: Rise of an Empire'

#### Finding the correlation
From the class reviews, for the movies I haven't seen, I would find the correlation of a movie I have seen and a movie I haven't seen in order to predict how I would rate them using pearson's r formula:
$$r =  \frac{\sum {(x-\overline{x})(y-\overline{y})}}{\sigma_x \sigma_y}$$

In [16]:
def pearson(s1, s2):
    """Take two pd.Series objects and return a pearson correlation."""
    s1_c = s1 - s1.mean()
    s2_c = s2 - s2.mean()
    return np.sum(s1_c * s2_c) / np.sqrt(np.sum(s1_c ** 2) * np.sum(s2_c ** 2))

### Correlation Between '300: Rise of an Empire' and movies I haven't seen
+ Gone Girl
+ The Fault in Our Stars
+ The Hobbit
+ The Lego Movie
+ Transformers
+ Unbroken

In [17]:
# Gone Girl
pearson(M['300: Rise of an Empire'],M['Gone Girl'])

0.071528915943837118

In [18]:
# The Lego Movie
pearson(M['300: Rise of an Empire'],M['The Fault in Our Stars'])

-0.060677823554797582

In [19]:
# The Hobbit
pearson(M['300: Rise of an Empire'],M['The Hobbit'])

0.068688456500578851

In [20]:
# The Fault in Our Stars
pearson(M['300: Rise of an Empire'],M['The Lego Movie'])

-0.09423609620781144

In [21]:
# Unbroken
pearson(M['300: Rise of an Empire'],M['Transformers'])

0.15431721047503769

In [22]:
# Tranformers
pearson(M['300: Rise of an Empire'],M['Unbroken'])

0.074996562323488239

### Correlation of 'Gone Girl' compared to all other Moview in the class

In [23]:
# Gone Girl
unseen_1 = get_recs('Gone Girl', M, 16)
unseen_1

[('American Sniper', 0.18971907113548794),
 ('How to Train your Dragon 2', 0.18031524427791323),
 ('The Lego Movie', 0.097308292851919775),
 ('Godzilla', 0.092140716054608607),
 ('Unbroken', 0.087777110648472254),
 ('Divergent', 0.08167993251956171),
 ('300: Rise of an Empire', 0.071528915943837118),
 ('The Fault in Our Stars', 0.05273965622202393),
 ('Interstellar', 0.028765721530187933),
 ('Guardians of the Galaxy', -0.036988810702596082),
 ('Transformers', -0.088695578823622639),
 ('Big Hero 6', -0.10574073070010113),
 ('The Hobbit', -0.13315814167705803),
 ('The Hunger Games: Mockingjay - Part 1', -0.13702913879535111),
 ('Malificent', -0.19520803666955933)]

In [24]:
# The Fault in Our Stars
unseen_2 = get_recs('The Fault in Our Stars', M, 16)
unseen_2

[('Malificent', 0.19749425916269375),
 ('Interstellar', 0.19712292369893003),
 ('Divergent', 0.14465573811263907),
 ('How to Train your Dragon 2', 0.14262156471653215),
 ('Guardians of the Galaxy', 0.13041089343894219),
 ('The Lego Movie', 0.083981926369952192),
 ('Unbroken', 0.06924931265846844),
 ('Gone Girl', 0.05273965622202393),
 ('The Hunger Games: Mockingjay - Part 1', 0.042352966198990802),
 ('Transformers', 0.007626920910922875),
 ('The Hobbit', -0.014562170243961364),
 ('Godzilla', -0.017978679468468192),
 ('300: Rise of an Empire', -0.060677823554797582),
 ('Big Hero 6', -0.083288293147851133),
 ('American Sniper', -0.087975224880283379)]

In [25]:
# The Hobbit
unseen_3 = get_recs('The Hobbit', M, 16)
unseen_3

[('Transformers', 0.29295673089206298),
 ('Guardians of the Galaxy', 0.21347311873933966),
 ('Malificent', 0.14457089779668711),
 ('Divergent', 0.13806981588918271),
 ('The Hunger Games: Mockingjay - Part 1', 0.13679981633621771),
 ('The Lego Movie', 0.13104564268551494),
 ('Godzilla', 0.10195136004390756),
 ('Interstellar', 0.088633055989652515),
 ('300: Rise of an Empire', 0.068688456500578851),
 ('How to Train your Dragon 2', 0.061870745688616106),
 ('The Fault in Our Stars', -0.014562170243961364),
 ('American Sniper', -0.061362767142021504),
 ('Big Hero 6', -0.06397997043606074),
 ('Gone Girl', -0.13315814167705803),
 ('Unbroken', -0.17101304297470835)]

In [26]:
# The Lego Movie
unseen_4 = get_recs('The Lego Movie', M, 16)
unseen_4

[('Guardians of the Galaxy', 0.40630056417316801),
 ('How to Train your Dragon 2', 0.22583645751392559),
 ('American Sniper', 0.18254986484255384),
 ('The Hobbit', 0.13104564268551494),
 ('Big Hero 6', 0.10569101716443542),
 ('Gone Girl', 0.097308292851919775),
 ('The Hunger Games: Mockingjay - Part 1', 0.095591088775192803),
 ('The Fault in Our Stars', 0.083981926369952192),
 ('Transformers', 0.043140080206105533),
 ('Godzilla', 0.041970869549338576),
 ('Interstellar', 0.033851976314638155),
 ('Divergent', 0.0023175740496638995),
 ('Malificent', -0.075474919195747267),
 ('300: Rise of an Empire', -0.09423609620781144),
 ('Unbroken', -0.13636935953252124)]

In [27]:
# Transformers
unseen_5 = get_recs('Transformers', M, 16)
unseen_5

[('Godzilla', 0.29963535888671367),
 ('The Hobbit', 0.29295673089206298),
 ('Interstellar', 0.20269967999375937),
 ('Divergent', 0.19305229322936257),
 ('The Hunger Games: Mockingjay - Part 1', 0.1682881078619495),
 ('300: Rise of an Empire', 0.15431721047503769),
 ('How to Train your Dragon 2', 0.15283445206468496),
 ('Malificent', 0.12899645848120822),
 ('Big Hero 6', 0.067131377670649625),
 ('The Lego Movie', 0.043140080206105533),
 ('Unbroken', 0.023254066831753972),
 ('American Sniper', 0.020929828825806115),
 ('The Fault in Our Stars', 0.007626920910922875),
 ('Guardians of the Galaxy', -0.044927356787947938),
 ('Gone Girl', -0.088695578823622639)]

In [28]:
# Unbroken
unseen_6 = get_recs('Unbroken', M, 16)
unseen_6

[('Divergent', 0.16612659382371495),
 ('Malificent', 0.14625631781343473),
 ('Gone Girl', 0.087777110648472254),
 ('Interstellar', 0.08493763888236304),
 ('Godzilla', 0.080353498077138555),
 ('300: Rise of an Empire', 0.074996562323488239),
 ('The Fault in Our Stars', 0.06924931265846844),
 ('Transformers', 0.023254066831753972),
 ('Big Hero 6', 0.010338098573443767),
 ('The Hunger Games: Mockingjay - Part 1', 0.0081914341546500315),
 ('How to Train your Dragon 2', -0.020864882870939724),
 ('Guardians of the Galaxy', -0.070334445276988428),
 ('The Lego Movie', -0.13636935953252124),
 ('American Sniper', -0.14355413959027591),
 ('The Hobbit', -0.17101304297470835)]

## The mean of all class reviews

In [29]:
M.describe()

C:\Users\millm_000\Anaconda3\lib\site-packages\numpy\lib\function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


American Sniper  The Hunger Games: Mockingjay - Part 1  \
count        73.000000                              83.000000   
mean          4.239726                               3.385542   
std           0.697762                               0.874173   
min           3.000000                               1.000000   
25%                NaN                                    NaN   
50%                NaN                                    NaN   
75%                NaN                                    NaN   
max           5.000000                               5.000000   

       Guardians of the Galaxy  The Lego Movie  The Hobbit  Transformers  \
count                95.000000       64.000000   99.000000    100.000000   
mean                  3.868421        3.406250    3.968687      3.580000   
std                   0.812831        1.079664    0.877223      1.158456   
min                   2.000000        1.000000    1.000000      1.000000   
25%                        NaN             NaN         NaN           NaN   
50%                        NaN             NaN         NaN           NaN   
75%                        NaN             NaN         NaN           NaN   
max                   5.000000        5.000000    5.000000      5.000000   

       Malificent  Big Hero 6   Godzilla  Interstellar  \
count   62.000000   55.000000  85.000000     85.000000   
mean     3.379032    4.023636   3.475294      4.564706   
std      0.903933    1.129235   0.976864      0.693457   
min      0.000000    1.000000   1.000000      2.000000   
25%           NaN         NaN        NaN           NaN   
50%           NaN         NaN        NaN           NaN   
75%           NaN         NaN        NaN           NaN   
max      5.000000    5.000000   5.000000      5.000000   

       How to Train your Dragon 2  Gone Girl  Divergent  \
count                   81.000000  65.000000  66.000000   
mean                     3.753086   3.953846   3.289394   
std                      0.905274   0.904516   0.813185   
min                      1.000000   2.000000   1.000000   
25%                           NaN        NaN        NaN   
50%                           NaN        NaN        NaN   
75%                           NaN        NaN        NaN   
max                      5.000000   5.000000   5.000000   

       The Fault in Our Stars   Unbroken  300: Rise of an Empire  
count               43.000000  31.000000               87.000000  
mean                 3.709302   3.661290                3.729885  
std                  0.920873   0.943284                1.150845  
min                  1.000000   2.000000                1.000000  
25%                       NaN        NaN                     NaN  
50%                       NaN        NaN                     NaN  
75%                       NaN        NaN                     NaN  
max                  5.000000   5.000000                5.000000

### Calculation of score prediction
Predicted Scores:
+ Gone Girl: 3.5
+ The Fault in Our Stars: 3.5
+ The Hobbit: 3.5
+ The Lego Movie: 3.5
+ Transformers: 3.5
+ Unbroken: 3.5

My calclations of my scores can be found at:
+ https://github.com/millmarc92/CSC-570---Data-Science/blob/master/M3/W12/Recommender_Q3.xlsx